### DISCUSSION TITLE
Athena / Redshift connection and assuming IAM role

### DISCUSSION THREAD

**Discussion by oscarxanda** [apache/superset#15739] [Q&A / Help] [answered]
Hello!

It is possible to connect to Athena / Redshift, and on the advanced tab, define the connect_args to use a IAM role. This allows to connect with the Key / Secret of account A, but then assume a role with specific permissions (set on the data lake).
The above works as expected.

However, I would like to have different roles for different users (so each user has its own set of permissions defined on the data lake). Would there be a way to achieve?

```
{
  "metadata_params": {},
  "engine_params": {
    "connect_args": {
      "role_arn": "arn:aws:iam::ACCOUNT_ID:role/ROLE_NAME"
    }
  },
  "metadata_cache_timeout": {},
  "schemas_allowed_for_csv_upload": [] 
}
```

**Comment by youcefjd** [comment]
Hi Oscar - can you explain how you made it work? I have the same config as you for Athena, but the connection doesn't go through.

**Reply by oscarxanda** [reply]
For my testing I ended up using the following db URI:
awsathena+rest://athena.eu-west-2.amazonaws.com/[DBNAME]?s3_staging_dir=[RESULTS_SAVE_LOCATION]
[RESULTS_SAVE_LOCATION] - must be uri encoded

Then on the extra meta parameters I had the following:
```
{
  "metadata_params": {},
  "engine_params": {
    "connect_args": {
"aws_access_key_id": "YOUR_KEY",
"aws_secret_access_key": "YOUR_SECRET_ACCESS",

      "role_arn": "ROLE_ARN"
    }
  },
  "metadata_cache_timeout": {},
  "schemas_allowed_for_csv_upload": [] 
}
```
The account that the YOUR_KEY and YOUR_SECRET_ACCESS pertains to should have permissions to call Athena.
The ROLE_ARN should have permissions to use athena and read / write to the buckets where data is stored / saved to.

**Comment by gforien** [comment, marked as answer]
Update on this answer because it helped me a lot ! Thank you [oscarxanda](https://github.com/oscarxanda)

I managed to make it work with this `datasource.yaml`
```
    - database_name: awsathena
      sqlalchemy_uri: awsathena+rest://athena.eu-west-2.amazonaws.com/DBNAME?s3_staging_dir=RESULTS_SAVE_LOCATION
      extra: "{\"engine_params\": {\"connect_args\": {\"role_arn\": \"{  ROLE_ARN  }\" }}}"
```

And the file `datasources.yaml` is imported at init with
`superset import_datasources -p /path/to/datasources.yaml`

Works like a charm, even with an IRSA role on AWS
